### PR TITLE
Add  ReadOptions implementation

### PR DESCRIPTION
**Pull request by y1zhou**
This is to address #125, where we pass in `ReadOptions` to the `open()` functions instead of just `StrictnessLevel` for future extensibility. 

- [x] PDB parser with `ReadOptions`
- [x] mmCIF parser with `ReadOptions`
- [x] Route to `open_gz` when `﻿options.decompress` is true
- [x] Remove hydrogens from the model when `options.discard_hydrogens` is true
- [x] Skip other models when `options.only_first_model` is true
- [x] Only parse `ATOM` records when `options.only_atomic_coords` is true

Some things up for discussion:

1. What is the intention of `﻿capitalise_chains`? If I recall correctly all uppercase letters, lowercase letters, and numbers are valid chain identifiers.
2. Can I modify the signatures of `open_pdb_raw` and `open_mmcif_raw`? They are public functions but it seems they are only meant to be called by `open_pdb` and `open_mmcif`. If we maintain `open_pdb_raw_with_options` and `open_mcif_raw_with_options` there's going to be a lot of duplicated code (or we extract common methods into another core function).

### CODE REVIEW

**Comment by douweschulte** [comment]
On your questions: 
1. This capitalises all chain ids before storing them, this was the standard behaviour before Oliver removed this behaviour.
2. I was thinking of making an `open_pdb_raw_with_options` as a new functions and then rewiring the original `open_pdb_raw` as a call to that function. If you think this would be too inconvenient for users we might think about breaking the API.

**Comment by OWissett** [comment]
I like the idea of keeping the API as simple as possible. I think that if we can have fewer functions for reading a file, that would be better, as I agree we don't want to duplicate code and also want it to be clear for end users to know which to use.

I see that there are two kind of ways to do this:

1. `open_pdb_with_opens(..., ReadOptions)`
2. Using the pattern from `OpenOptions` in the `std` library: https://doc.rust-lang.org/std/fs/struct.OpenOptions.html

I was initially wanting to change it to be like option 2, but from a time perspective, I just needed to add the functionality as fast as I could, so I went with 1. 

What do you think>

**Comment by douweschulte** [comment]
I think that the open options would be the more clear way of doing this and allowing further expansion in the future. So if everyone here agree I think we can make this a breaking change and move to make this the default. After this work has been done it might be worthwhile to see if we can trivially add back the original public functions with the same api and mark them as [deprecated](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-deprecated-attribute) to be removed in the next breaking change (`v0.13` ) so that users have at least one major version to move their stuff over.

**Comment by y1zhou** [comment]
Yeah I'm already down the rabbit hole of `open_*_with_options()` so let's try that first. I also like the idea of eventually deprecating the functions without `ReadOptions` to avoid confusion.

**Comment by OWissett** [comment]
Sounds good. I theory, we are able to have both the approaches of 1 and 2. 

I like using the builder pattern, as it provides a good extensible framework where we can add more options without changing the API, this applies to both approaches, but I think that 2 is more idiomatic

**Comment by y1zhou** [comment]
Alright most implementations are done with the following exceptions:

1. Atom-only mode for mmCIF files - I'm not familiar enough with the CIF format to know which fields could be safely skipped.
2. `ReadOptions.format`: do we keep this for route 2? Currently the input file type is determined at many different places and it might be nice to have a single source of truth.
3. `ReadOptions.capitalise_chains`: seems like it's being used in `chain_iter` of `open_pdb_raw_with_options` as a fallback? Not sure how the option should interact with this during parsing. Also for mmCIF files there doesn't seem to be a counterpart for this.

File: tests/ignore_hydrogens.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// Test reading PDB file and discarding all hydrogen atoms.
+
+use pdbtbx::*;
+
+#[test]
+fn main() {
```
Comment (OWissett) [inline tests/ignore_hydrogens.rs:6]: can you give the test a more descriptive name?

Comment (y1zhou) [reply tests/ignore_hydrogens.rs]: Done!

**Review by douweschulte — COMMENTED** [review]
Looks really great! It looks like you made a lot of new public methods for opening files. It looks like we all agree on having the `ReadOptions` be the main way of opening files moving forwards so it would be better to mark these functions pub(crate) or not export them in the `mod.rs` file. On that same vein I think it would be best to add a deprecated warning to the functions that will still need to be exported. Otherwise I am impressed by your Rust code, looks really good and idiomatic!

**Comment by y1zhou** [comment]
> It looks like you made a lot of new public methods for opening files.

Now is about time to decide which pub functions to keep and which ones to remove (or deprecate). Here's my two cents:

## General

Only keep `open(filename)` and `open_with_options(filename, options)`.

- `open` would be `open_with_options` with sane default options. Now that `level` lives in `ReadOptions` I would also suggest removing the `level` argument in `open()`.
- `open_gz` can be called by setting `options.decompress` in `open_with_options`.
- `open_raw` does some extra magic in filetype guessing which is not very reliable, e.g. an atom-only PDB file doesn't start with `HEADER` and cannot be identified (somewhat related #122).
  - Users could use `open_pdb_raw_with_options` and `open_mmcif_raw_with_options` directly.
  - Alternatively, we utilize `ReadOptions.format` and let this determine which internal function to route to. 

## PDB and mmCIF

- Keep `open_pdb(filename)`, `open_pdb_with_options(filename, options)`, `open_pdb_raw_with_options(BufReader, context, options)`.
- Keep `open_mmcif(filename)`, `open_mmcif_with_options(filename, options)`, and `open_mmcif_raw_with_options(BufReader, context, options)`.

On the `open_*_raw` functions, currently `open_pdb_raw` takes `BufReader` but `open_mmcif_raw` takes `&str`. I think it makes more sense to just expose a function that takes `BufReader` as the input can be streamed, and converting a `&str` in memory to a `BufReader` is trivial.

To summarize,

| Function | Action |
|----------|-------|
| `general::open` | Remove `level` argument |
| `general::open_gz` | Deprecate |
| `general::open_raw` | Deprecate |
| `pdb::open_pdb` | Remove `level` argument |
| `pdb::open_pdb_raw` | Deprecate |
| `mmcif::open_mmcif` | Remove `level` argument |
| `mmcif::open_mmcif_bufread` | Deprecate |
| `mmcif::open_mmcif_bufread_with_options` | Deprecate |
| `mmcif::open_mmcif_raw` | Deprecate |
| `mmcif::open_mmcif_raw_with_options` | Take `BufReader` instead of `&str`, and parse line-by-line instead |

**Comment by OWissett** [comment]
I like the sound of those changes.

I agree that we still want to have a public function which takes a `BufReader` as an input, since people may want to do some more complex tasks than getting the library to handling the opening of a file, e.g., maybe receiving over a socket or something...

**Comment by douweschulte** [comment]
These changes sound very good indeed. But how does this look on the `ReadOptions` side? There is a function `read` which takes an `&str` right now. I would argue that adding a `ReadOptions::read_raw` is enough to gain feature parity between the old system and the new system. If that function is added I would say we just expose the `ReadOptions` functions. But to ease the development for all users I think exposing the exact same api as the old system makes sense, but all of these functions should be marked `deprecated`.

**Comment by y1zhou** [comment]
Sounds good. I'll add the functions to `ReadOptions` as well and see how it looks.

**Comment by y1zhou** [comment]
Things might be a little messy but the changes we discussed last time should be updated now. Please let me know if I missed anything!

File: src/read/read_options.rs [hunk 2]
```diff
@@ -81,6 +84,28 @@ impl ReadOptions {
         self
     }
 
+    /// Guess the file format based on the file name extensions.
+    pub fn guess_format(&mut self, filename: &str) -> &mut Self {
+        let extensions: Vec<&str> = filename.split('.').collect();
+        match extensions[extensions.len() - 1] {
```
Comment (douweschulte) [inline src/read/read_options.rs:90]: This will crash on files without any extension. It might be worthwhile to look into using [`Path::extension`](https://doc.rust-lang.org/std/path/struct.Path.html#method.extension) to not have to think about all edge cases that exist.

File: src/read/read_options.rs [hunk 3]
```diff
@@ -112,8 +137,94 @@ impl ReadOptions {
...
+    /// The correct type will be determined based on the file extension.
+    ///
+    /// # Errors
+    /// Returns a `PDBError` if a `BreakingError` is found. Otherwise it returns the PDB with all errors/warnings found while parsing it.
+    ///
+    /// # Related
+    /// If you want to open a file from memory see [`ReadOptions::read_raw`].
+    pub fn read(&self, path: impl AsRef<str>) -> ReadResult {
+        if self.decompress {
+            // open a decompression stream
```
Comment (douweschulte) [inline src/read/read_options.rs:156]: Do I read the code correctly that the decompress branch does not auto detect the file type? Here the filename is still known so it could do the auto detect functionality.

**Review by douweschulte — COMMENTED** [review]
Good work, two small comments left from my side.

File: src/read/read_options.rs (hunk 2 above)
Comment (douweschulte) [reply src/read/read_options.rs:90]: Also I think I saw this same logic match in multiple places (sometimes with and sometimes without the compression) it might be good to refactor this into one function and call that from all places where the logic is needed.

File: src/read/read_options.rs (hunk 3 above)
Comment (y1zhou) [reply src/read/read_options.rs:144]: You are absolutely right! I was hoping users to call `self.guess_format` explicitly before calling `read`, otherwise `read` would need a mutable reference of `self` since `self.format` would be modified when guessing the format from the file name. Do you have recommendations for handling this?

Comment (douweschulte) [reply src/read/read_options.rs:144]: I think that having a generalised function that takes a path and returns a `Format` (or something that also included the compression when needed) would solve your problem. The `self.guess_format` then calls that and updates itself while any place can just use the result of the generalised function locally and does not need to store anything per se.

File: src/read/read_options.rs [hunk 4]
```diff
@@ -112,8 +126,128 @@ impl ReadOptions {
...
+                "Could not determine file type",
+                "Could not determine the type of the input stream, set self.format",
+                Context::None,
+            )]),
+        }
+    }
+}
+
+/// Guess the file format based on the file name extensions.
+fn guess_format(filename: &str) -> Option<(Format, bool)> {
```
Comment (y1zhou) [inline src/read/read_options.rs:237]: @douweschulte Is this what you suggested in the previous comment? I believe now errors are only raised when `read_raw` is called directly on a stream and `options.format` is not set. In all other scenarios the file type should be inferred from the filename.

Comment (douweschulte) [reply src/read/read_options.rs:237]: Looks good!

**Review by douweschulte — APPROVED** [review]

File: src/read/pdb/parser.rs [hunk 5]
```diff
@@ -18,9 +19,17 @@ use std::io::{BufRead, BufReader};
 /// # Related
 /// If you want to open a file from memory see [`open_pdb_raw`]. There is also a function to open a mmCIF file directly
```
Comment (OWissett) [inline src/read/pdb/parser.rs:20]: This comment needs updating too, currently it points to a function marked as deprecated

File: src/read/pdb/parser.rs [hunk 6]
```diff
@@ -18,9 +19,17 @@ use std::io::{BufRead, BufReader};
...
-pub fn open_pdb(
+pub fn open_pdb(filename: impl AsRef<str>) -> Result<(PDB, Vec<PDBError>), Vec<PDBError>> {
+    open_pdb_with_options(filename, &ReadOptions::default())
+}
+
+/// Parse the given file into a PDB struct with [`ReadOptions`].
+///
+/// # Related
+/// See [`open_pdb`] for a version of this function with sane defaults.
+pub fn open_pdb_with_options(
```
Comment (OWissett) [inline src/read/pdb/parser.rs:30]: Should this be made `pub(crate)`? Since we want people to use a single API, through methods on the `ReadOptions` struct

File: src/read/pdb/parser.rs [hunk 7]
```diff
@@ -45,11 +54,30 @@ pub fn open_pdb(
...
+    T: std::io::Read,
+{
+    open_pdb_raw_with_options(input, context, ReadOptions::default().set_level(level))
+}
+
+/// Parse the input stream into a PDB struct with [`ReadOptions`].
+///
+/// # Related
+/// See [`open_pdb_raw`] for a version of this function with sane defaults.
+pub fn open_pdb_raw_with_options<T>(
```
Comment (OWissett) [inline src/read/pdb/parser.rs:76]: also `pub(crate)`

File: src/read/read_options.rs [hunk 8]
```diff
@@ -1,6 +1,6 @@
-use crate::StrictnessLevel;
+use crate::{Context, PDBError, StrictnessLevel};
 
-use super::general::{open_with_options, ReadResult};
+use super::general::ReadResult;
```
Comment (OWissett) [inline src/read/read_options.rs:5]: Is this instead of doing `pub(crate)`? I think I saw a conversation about this. In my opinion, it would be better to simply completely hide the interface, as it means we have more freedom later to change things without affecting the API (which is the may goal of using the options pattern)

File: src/read/pdb/parser.rs (hunk 7 above)
Comment (OWissett) [reply src/read/pdb/parser.rs:76]: Alternatively, we mark this function as `#[unstable]` so that people can't simply use them and not realise that they may change

File: src/structs/pdb.rs [hunk 9]
```diff
@@ -468,7 +465,7 @@ impl<'a> PDB {
     /// Find all hierarchies matching the given search. For more details see [Search].
     /// ```
     /// use pdbtbx::*;
-    /// let (pdb, errors) = open_pdb("example-pdbs/1ubq.pdb", StrictnessLevel::Loose).unwrap();
+    /// let (pdb, errors) = open_pdb_with_options("example-pdbs/1ubq.pdb", ReadOptions::new().set_level(StrictnessLevel::Loose)).unwrap();
```
Comment (OWissett) [inline src/structs/pdb.rs:468]: Change this from `open_pdb_with_options` to the `ReadOptions::read()`

Comment (OWissett) [reply src/structs/pdb.rs:468]: or just make it use `open` assuming that the file is ok with the default strictness

**Review by OWissett — CHANGES_REQUESTED** [review]

File: src/read/pdb/parser.rs (hunk 6 above)
Comment (y1zhou) [reply src/read/pdb/parser.rs:30]: Marking these functions as deprecated in this release and making them `pub(crate)` might be easier for people to make the transition? Making them `pub(crate)` now would break all existing code.

Comment (OWissett) [reply src/read/pdb/parser.rs:30]: Ok cool. Mark as deprecated

File: src/read/read_options.rs (hunk 8 above)
Comment (OWissett) [reply src/read/read_options.rs:5]: @douweschulte Thoughts?

I would be in favour of only having `open` and `ReadOptions`, and getting rid of `open_with_options` in the public API, either deleting or marking as `pub(crate)`

Comment (y1zhou) [reply src/read/read_options.rs:5]: > @douweschulte Thoughts?
> 
> I would be in favour of only having `open` and `ReadOptions`, and getting rid of `open_with_options` in the public API, either deleting or marking as `pub(crate)`

I think this makes sense, as `open` would be the "magic" function with sane defaults, and `ReadOptions` would be the only API used for fine-grained controls. We may even take a step further and mark `open_*`, `open_*_with_options`, and `open_*_raw` as deprecated and steer users towards `ReadOptions` in these cases.

Comment (douweschulte) [reply src/read/read_options.rs:5]: I fully agree. In the end we want to only have the `ReadOptions` and potentially `open` functions. Any function added by this PR can be marked `pub(crate)` any function already present in the previous version has to be marked `deprecated` for at least one major version to help users migrate.

Comment (OWissett) [reply src/read/read_options.rs:5]: I like the sound of that.

Comment (y1zhou) [reply src/read/read_options.rs:5]: Should all be deprecated or marked `pub(crate)` now. Please check.

**Comment by douweschulte** [comment]
To help you out this is the output from [`cargo-public-api`](https://github.com/enselic/cargo-public-api) run as follows: `cargo public-api diff 0.11.0` on your feature branch:
```
Removed items from the public API
=================================
-pub fn pdbtbx::open_raw<T: std::io::Read + std::io::Seek>(input: std::io::buffered::bufreader::BufReader<T>, level: pdbtbx::StrictnessLevel) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>

Changed items in the public API
===============================
-pub fn pdbtbx::open(filename: impl core::convert::AsRef<str>, level: pdbtbx::StrictnessLevel) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>
+pub fn pdbtbx::open(filename: impl core::convert::AsRef<str>) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>
-pub fn pdbtbx::open_mmcif(filename: impl core::convert::AsRef<str>, level: pdbtbx::StrictnessLevel) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>
+pub fn pdbtbx::open_mmcif(filename: impl core::convert::AsRef<str>) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>
-pub fn pdbtbx::open_pdb(filename: impl core::convert::AsRef<str>, level: pdbtbx::StrictnessLevel) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>
+pub fn pdbtbx::open_pdb(filename: impl core::convert::AsRef<str>) -> core::result::Result<(pdbtbx::PDB, alloc::vec::Vec<pdbtbx::PDBError>), alloc::vec::Vec<pdbtbx::PDBError>>

Added items to the public API
=============================
(ignored)
```

For full API compatibility there should be no functions listed in the removed and changed items section. So in this case that means keeping the `open_raw` and keeping the `strictnesslevel` parameter in the other open functions. As discussed all open functions we do not intend to support should be marked `deprecated`.

It looks like the original `open_raw` definition is not easily reproduced with the `ReadOptions` as it determined the file type with the first characters being `HEADER` or `data_`. So if it is not possible to keep this function we could drop this one already in this release.

**Comment by y1zhou** [comment]
> To help you out this is the output from [`cargo-public-api`](https://github.com/enselic/cargo-public-api) run as follows: `cargo public-api diff 0.11.0` on your feature branch:

This is pretty handy! Thanks for the pointers.

> For full API compatibility there should be no functions listed in the removed and changed items section. So in this case that means keeping the `open_raw` and keeping the `strictnesslevel` parameter in the other open functions. As discussed all open functions we do not intend to support should be marked `deprecated`.
> 
> It looks like the original `open_raw` definition is not easily reproduced with the `ReadOptions` as it determined the file type with the first characters being `HEADER` or `data_`. So if it is not possible to keep this function we could drop this one already in this release.

I agree that `open_raw` should be dropped and marked as a breaking change. Using the first few characters to determine the file type becomes less reliable especially with the `atomic_only` modifiers being introduced.

For the other functions, adding back `level` to `open_pdb` and `open_mmcif` makes sense as they are deprecated anyway. For `open`, what would be the best approach to modify its signatures? In our previous discussions I think we agreed to let `open` use all default options, and `ReadOptions` to handle all other cases. Do we still want the old `open(filename, level)` or just `open(filename)`?

**Comment by douweschulte** [comment]
For `open` lets look at the standard library. There there are very simple 'magical' functions alongside the open options, which is exactly as discussed before. So we should go for that (using `open(filename)`) and we should try and make it abundantly clear how to convert existing code in the release notes.

**Comment by y1zhou** [comment]
Sounds good! We have one removed item `open_raw` and one changed item `open`. Hopefully it's not too much burden to the end users to migrate existing code to `ReadOptions`. Looking forward to v0.12.0!

**Comment by y1zhou** [comment]
Just want to follow up on this PR and see if there's anything to be updated before this can be merged.

**Comment by douweschulte** [comment]
I propose a merge (I was awaiting a response from @OWissett).

**Comment by OWissett** [comment]
thanks for merging. Apologies for delay, I was on holiday
